File: Dynamic Programming/Min cost_path_sum/minimum_path_sum_LC64.go
func minPathSum(grid [][]int) int {
    rows := len(grid)
    if rows == 0 {
        return 0
    }
    cols := len(grid[0])


    for i:=rows-1; i>=0; i-- {
        for j:= cols-1; j>=0; j-- {
            if (i==rows-1 && j==cols-1) {
                continue
            } else if (i== rows-1){
                grid[i][j] += grid[i][j+1]
            } else if (j == cols-1){
                grid[i][j] += grid[i+1][j]
            } else {
                grid[i][j] += min(grid[i+1][j], grid[i][j+1])
            }
        }
    }
    return grid[0][0]
}

func min(a, b int) int{
    if a<b {
        return a
    }
    return b
}

File: Dynamic Programming/Min cost_path_sum/minimum_path_sum_LC64.py
# Given a m x n grid filled with non-negative numbers, find a path from top left to bottom right, which minimizes the sum of all numbers along its path.
# Note: You can only move either down or right at any point in time.

class Solution(object):
    def minPathSum(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """

        m = len(grid)
        n = len(grid[0]) if m else 0

        for i in range(m-1,-1,-1):
            for j in range(n-1,-1,-1):
                if i == m-1 and j == n-1:
                    pass
                elif i == m-1:
                    grid[i][j] += grid[i][j+1]
                elif j == n-1:
                    grid[i][j] += grid[i+1][j]
                else:
                    #print(i,j)
                    grid[i][j] += min(grid[i+1][j], grid[i][j+1])
        return grid[0][0]
